From: Chen Wandun <chenwandun@huawei.com>
Subject: mm/migrate: continue to migrate for non-hugetlb folios
Date: Sun, 29 Jan 2023 11:39:10 +0800

migrate_hugetlbs() returns -ENOMEM when there are not enough hugetlbs,
however there may be free non-hugetlb folios available, so continue to
migrate for non-hugetlb folios.

Nowdays the only negative return value of migrate_hugetlbs is -ENOMEM. 
Keep compatibility for coming negative return values instead of ignore
the return value.

Link: https://lkml.kernel.org/r/20230129033910.1327277-1-chenwandun@huawei.com
Signed-off-by: Chen Wandun <chenwandun@huawei.com>
Reviewed-by: "Huang, Ying" <ying.huang@intel.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---


--- a/mm/migrate.c~mm-migrate-continue-to-migrate-for-non-hugetlb-folios
+++ a/mm/migrate.c
@@ -1939,7 +1939,8 @@ int migrate_pages(struct list_head *from
 
 	rc_gather = migrate_hugetlbs(from, get_new_page, put_new_page, private,
 				     mode, reason, &stats, &ret_folios);
-	if (rc_gather < 0)
+	/* There may be free non-hugetlb folios available, continue to migrate. */
+	if (rc_gather < 0 && rc_gather != -ENOMEM)
 		goto out;
 again:
 	nr_pages = 0;
_
